### PR TITLE
server/ui: fix crash caused by race in server initialization

### LINKED ISSUE #25771: ui: race condition during initialization with login enabled
When login is enabled, there seems to be a race condition during initialization, manifesting as the error "settings/cluster/settings.go:164  Version() was called before having been initialized".  It seems to occur when an existing admin ui browser tab is open during node startup, issuing requests such as health checks.  It was previously theorized that #24945 fixed this issue, but it looks like it's come up since.

Stack trace for my repro just now:

```
$ COCKROACH_EXPERIMENTAL_REQUIRE_WEB_LOGIN=true ./cockroach start --certs-dir clusters/login-test/certs --store clusters/login-test/cockroach-data/
F180521 18:20:17.173359 127 settings/cluster/settings.go:164  Version() was called before having been initialized
goroutine 127 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.getStacks(0xc4204fa500, 0xc4204fa540, 0x6fe6c00, 0x1c)
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:956 +0xcf
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).outputLogEntry(0x76ab5e0, 0xc400000004, 0x6fe6cb6, 0x1c, 0xa4, 0xc4207f6400, 0x33)
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:832 +0x753
github.com/cockroachdb/cockroach/pkg/util/log.addStructured(0x63f4aa0, 0xc420054098, 0x4, 0x2, 0x0, 0x0, 0xc420b96c08, 0x1, 0x1)
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/util/log/structured.go:154 +0x2e5
github.com/cockroachdb/cockroach/pkg/util/log.logDepth(0x63f4aa0, 0xc420054098, 0x1, 0xc400000004, 0x0, 0x0, 0xc420b96c08, 0x1, 0x1)
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:55 +0x8c
github.com/cockroachdb/cockroach/pkg/util/log.Fatal(0x63f4aa0, 0xc420054098, 0xc420b96c08, 0x1, 0x1)
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:158 +0x6c
github.com/cockroachdb/cockroach/pkg/settings/cluster.(*ExposedClusterVersion).Version(0xc420705830, 0x0, 0x0, 0x0, 0x5ea7e60)
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/settings/cluster/settings.go:164 +0x161
github.com/cockroachdb/cockroach/pkg/settings/cluster.(*ExposedClusterVersion).IsActive(0xc420705830, 0x7, 0x608d904)
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/settings/cluster/settings.go:199 +0x2b
github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).performRangeLookup(0xc42052a420, 0x63f4b20, 0xc42024f260, 0xc4202b1c90, 0xe, 0x10, 0x6417800, 0xc4202868c0, 0xc4206eb000, 0x1000, ...)
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:419 +0x103
github.com/cockroachdb/cockroach/pkg/kv.(*RangeDescriptorCache).lookupRangeDescriptorInternal.func3(0x0, 0x0, 0x0, 0x0)
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/kv/range_cache.go:310 +0x132
github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).doCall(0xc42052a460, 0xc42025e780, 0xc420312820, 0x14, 0xc4206b7d00)
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:118 +0x2e
created by github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight.(*Group).DoChan
	/Users/couch/go/src/github.com/cockroachdb/cockroach/pkg/util/syncutil/singleflight/singleflight.go:111 +0x2d0
```

### PR DESCRIPTION
**Pull request by vilterp**
Fixes #25771

See commits for details.

### CODE REVIEW

**Review by benesch — APPROVED** [review]
The commits you wrote :lgtm_strong:. The commit I wrote was LGTM'd by Peter in #30735. So I think we're good to go here, modulo one stray println I called out below. Thanks for indulging me on this approach.

> Reviewed 1 of 1 files at r1, 4 of 4 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30746#-:-LN_1uepEZFgeDc39w3e:b-t728xh)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/server/server.go, line 1583 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/30746#-LNW_LP7E2utfjauD8Am:-LNW_LP7E2utfjauD8An:b-vafdme) ([raw file](https://github.com/cockroachdb/cockroach/blob/5be327520f72a63f4bc8afda117df9452e7ca6cd/pkg/server/server.go#L1583)):*
> ```Go
> 
> 	// Serve UI assets. This needs to be before the gRPC handlers are registered, otherwise
> 	// the `s.mux.Handle("/", ...)` would cover all URLs, allowing anonymous access.
> ```

I know you're just copy/pasting, but I don't understand what this paragraph is saying. And whatever it's saying I'm not sure is true? I don't see why the order in which the routes are installed would matter.
___
*[pkg/ui/ui.go, line 127 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/30746#-LNWYt7yBS4GPDrtbP1N:-LNWYt7yBS4GPDrtbP1O:b-7j8trw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5be327520f72a63f4bc8afda117df9452e7ca6cd/pkg/ui/ui.go#L127)):*
> ```Go
> 
> 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
> 		fmt.Println("hello from ")
> ```

stray

**Review by couchand — APPROVED** [review]
LGTM, thanks for getting to the bottom of this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30746#-:-LNkJ1rK9XXKO69RZJHl:bk8tol)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Review by couchand — APPROVED** [review]
> Reviewed 3 of 4 files at r2, 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30746)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Review by vilterp — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/30746)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/server/server.go, line 1583 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/30746#-LNW_LP7E2utfjauD8Am:-LNkQ-Zq7fVRXspEVjI2:b-ve0diy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5be327520f72a63f4bc8afda117df9452e7ca6cd/pkg/server/server.go#L1583)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I know you're just copy/pasting, but I don't understand what this paragraph is saying. And whatever it's saying I'm not sure is true? I don't see why the order in which the routes are installed would matter.
</blockquote></details>

Removed. The order can matter when there is some time between registrations (as there used to be in this method) but that's not really what the comment is explaining. Not quite sure what I was thinking when I wrote it.
___
*[pkg/ui/ui.go, line 127 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/30746#-LNWYt7yBS4GPDrtbP1N:-LNkP8Rh3aC8dknEhZep:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5be327520f72a63f4bc8afda117df9452e7ca6cd/pkg/ui/ui.go#L127)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

stray
</blockquote></details>

Done.

File: pkg/server/server.go [hunk 1]
```diff
@@ -1584,9 +1579,13 @@ func (s *Server) Start(ctx context.Context) error {
...
-	// Serve UI assets. This needs to be before the gRPC handlers are registered, otherwise
-	// the `s.mux.Handle("/", ...)` would cover all URLs, allowing anonymous access.
-	maybeAuthMux := newAuthenticationMuxAllowAnonymous(
+	// Serve UI assets.
+	//
+	// The authentication mux used here is created in "allow anonymous" mode so that the UI
+	// assets are served up whether or not there is a session. If there is a session, the mux
+	// adds it to the context, and it is templated into index.html so that the UI can show
+	// the username of the currently-logged-in user.
+	authenticatedUIHandler := newAuthenticationMuxAllowAnonymous(
```
Comment (couchand) [inline pkg/server/server.go]: nit: I think `mux` is more accurate than `handler` for this, since it's not handling any requests itself.

Comment (vilterp) [reply pkg/server/server.go]: it's a handler wrapped in an auth mux (which fulfills the `http.Handler`) interface, which is passed to `s.mux.Handle(path, http.Handler)`… so `¯\_(ツ)_/¯`

**Pull request merged**